### PR TITLE
Reviewer Rob - Common C++ build infrastructure

### PR DESCRIPTION
**Pull request by bossmc**
This adds a common C++ build infrastructure makefile snippet which manages:

* Building release binaries
* Building test binaries
* Running tests
* Running coverage
* Running valgrind
* Managing build dependencies

### CODE REVIEW

**Comment by bossmc** [comment]
See https://github.com/Metaswitch/ralf/pull/191 for an example of how to use this snippet.

File: cpp.mk [hunk 1]
```diff
@@ -0,0 +1,181 @@
+# Common C++ Makefile infrastructure
+#
+# Product specific Makefiles define a list of TARGETS and TEST_TARGETS
+# and then `include` this Makefile.  To customise the build product
+# Makefiles can define the following to tweak the behavior:
+#
+#   target_SOURCES  - List the .cpp files to build (defaults to target.cpp only)
```
Comment (rkd-msw) [inline cpp.mk]: Could we say `<target>` to ensure it's clear that it's a placeholder?

File: cpp.mk [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+# Makefiles can override these if needed
+GTEST_DIR ?= ../modules/gmock/gtest
+GMOCK_DIR ?= ../modules/gmock
+CPP_COMMON_DIR ?= ../modules/cpp-common
+BUILD_DIR ?= ../build
+
+# Common rules to build any target
+#
+# @param $1 - Target name (final executable)
+# @param $2 - Target
```
Comment (rkd-msw) [inline cpp.mk]: What's the difference between these?

File: cpp.mk [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+define common_target
+
+# Calculate the object names for the $1 build
+$1_OBJS := $$(patsubst %.cpp,$${BUILD_DIR}/$1/%.o,$${$1_SOURCES})
+$1_DEPS := $$(patsubst %.cpp,$${BUILD_DIR}/$1/%.d,$${$1_SOURCES})
+
+# Create alias for the object directory (for adding custom pre-requisites)
```
Comment (rkd-msw) [inline cpp.mk]: I don't understand how this is useful "for adding custom pre-requisites".

File: cpp.mk [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+$1_OBJECT_DIR := $${BUILD_DIR}/$1
+
+# Depend on the depends file so we'll get rebuilt if it's missing
+# (which will have the side effect of re-producing the depends file)
+$${$1_OBJS} : $${$1_OBJECT_DIR}/%.o : %.cpp $${BUILD_DIR}/$1/%.d ${MAKEFILE_LIST}
```
Comment (rkd-msw) [inline cpp.mk]: What's this syntax with two colons?

Comment (rkd-msw) [inline cpp.mk]: Maybe worth a link to something that explains the `%.o` for-all-X syntax?

File: cpp.mk [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+	@mkdir -p $${$1_OBJECT_DIR}
+	g++ -MMD -MP $${$2_CPPFLAGS} $${$1_CPPFLAGS} -c $$< -o $$@
+
+# Blank rule for depends files to force the object to be re-built
```
Comment (rkd-msw) [inline cpp.mk]: Is this redundant? The rule above says it accomplishes the same thing.

File: cpp.mk [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+GTEST_DIR ?= ../modules/gmock/gtest
+GMOCK_DIR ?= ../modules/gmock
+CPP_COMMON_DIR ?= ../modules/cpp-common
+BUILD_DIR ?= ../build
+
+# Common rules to build any target
+#
+# @param $1 - Target name (final executable)
+# @param $2 - Target
+define common_target
```
Comment (rkd-msw) [inline cpp.mk]: I hadn't seen `define` before and didn't realise it started a block - comment?

File: cpp.mk [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+		sort > $${BUILD_DIR}/scratch/valgrind.tmp
+	@if grep -q -v "XPath set is empty" $${BUILD_DIR}/scratch/valgrind.tmp ; then \
+		echo "Error: some memory errors have been detected" ; \
+		cat $${BUILD_DIR}/scratch/valgrind.tmp ; \
+		echo "See $$< for further details." ; \
+		exit 2 ; \
+	fi
+
+.PHONY : valgrind_$1
+valgrind_$1 : $${BUILD_DIR}/bin/$1
```
Comment (rkd-msw) [inline cpp.mk]: I think this replaces vg_raw - is that right?

File: cpp.mk [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+		cat $${BUILD_DIR}/scratch/coverage_$1_filtered.tmp ; \
+		exit 2 ; \
+	fi
+CLEANS += $${BUILD_DIR}/scratch/coverage_$1.tmp $${BUILD_DIR}/scratch/coverage_$1_filtered.tmp
+
+.PHONY : coverage_raw_$1
+coverage_raw_$1 : $${BUILD_DIR}/$1/.$1_already_run
+	@../modules/gcovr/scripts/gcovr $${$1_COVERAGE_ARGS} --keep --sort-percentage
+
+test : run_$1
```
Comment (rkd-msw) [inline cpp.mk]: `make test` should do everything - it should run the tests, run them under valgrind, and check coverage.

File: cpp.mk [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+$(call common_target,$1,test)
+
+$${BUILD_DIR}/bin/$1 : $${BUILD_DIR}/obj/gmock-all.o $${BUILD_DIR}/obj/gtest-all.o
+
+.PHONY : run_$1
+run_$1 : $${BUILD_DIR}/bin/$1
+	LD_LIBRARY_PATH=../usr/lib/ $$< ${EXTRA_TEST_ARGS}
+
+# This sentinel file proves the tests have *all* been run on this build (mostly for coverage)
+$${BUILD_DIR}/$1/.$1_already_run : $${BUILD_DIR}/bin/$1
```
Comment (rkd-msw) [inline cpp.mk]: Does this mean you have to run the tests twice when `make test` runs?

File: cpp.mk [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+vg_check : valgrind_check_$1
+coverage_check : coverage_check_$1
+coverage_raw : coverage_raw_$1
+
+CLEANS += $${BUILD_DIR}/$1/valgrind_output.xml $${BUILD_DIR}/$1/.$1_already_run
+
+endef
+
+# Default values for build flags for each build
+__COMMON_CPPFLAGS := -ggdb3 -std=c++11
```
Comment (rkd-msw) [inline cpp.mk]: -Wall?

Comment (rkd-msw) [inline cpp.mk]: Also, nice try at sneaking it through, but we're `-std=c++0x` at the moment so this change at least warrants a team heads-up.

File: cpp.mk [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+$(foreach target,${TARGETS},$(eval $(call common_target,${target},release)))
+$(foreach target,${TEST_TARGETS},$(eval $(call test_target,${target})))
+
+# Build all non-test targets by default
+all : ${TARGETS}
+
+clean :
+	-rm $(sort ${CLEANS}) # make's sort function removes duplicates as a side effect
+
+print-% :
```
Comment (rkd-msw) [inline cpp.mk]: What?

File: cpp.mk [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+$${BUILD_DIR}/$1/.$1_already_run : $${BUILD_DIR}/bin/$1
+	LD_LIBRARY_PATH=../usr/lib/ $$<
+	@touch $$@
+
+.PHONY : debug_$1
+debug_$1 : $${BUILD_DIR}/bin/$1
+	LD_LIBRARY_PATH=../usr/lib/ gdb --args $$< $${EXTRA_TEST_ARGS}
+
+$${BUILD_DIR}/$1/valgrind_output.xml : $${BUILD_DIR}/bin/$1
+	LD_LIBRARY_PATH=../usr/lib/ valgrind --xml=yes --xml-file=$$@ $$<
```
Comment (rkd-msw) [inline cpp.mk]: Ralf's old makefile had:

```VGFLAGS = --suppressions=$(VG_SUPPRESS) \
          --gen-suppressions=all \
          --leak-check=full \
          --track-origins=yes \
          --malloc-fill=cc \
          --free-fill=df
```

and `VG_SUPPRESS = ut/$(TARGET_TEST).supp`. What's the new equivalent?

File: cpp.mk [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+coverage_check : coverage_check_$1
+coverage_raw : coverage_raw_$1
+
+CLEANS += $${BUILD_DIR}/$1/valgrind_output.xml $${BUILD_DIR}/$1/.$1_already_run
+
+endef
+
+# Default values for build flags for each build
+__COMMON_CPPFLAGS := -ggdb3 -std=c++11
+release_CPPFLAGS := -O2 ${__COMMON_CPPFLAGS}
```
Comment (rkd-msw) [inline cpp.mk]: Have we lost the "filter out other optimization levels" thing?

**Comment by rkd-msw** [comment]
Looks OK. Is there anything you've deliberately changed in the port which I should specifically consider?

File: cpp.mk (hunk 12 above)
Comment (bossmc) [inline cpp.mk]: Good spot, I've moved most of that to common and Ralf's Makefile now (optionally) provides a  custom suppression file.

File: cpp.mk (hunk 7 above)
Comment (bossmc) [inline cpp.mk]: Yes, expect a PSA for this (and the few other changes)

File: cpp.mk (hunk 8 above)
Comment (bossmc) [inline cpp.mk]: Maybe, I think it's probably useful to have a target that runs all the tests and one that runs all the tests slowly as a final check (in particular, CI can use the full test suite) for developers before they merge a PR.

I don't have a good naming proposal for these two targets though (`make test` and `make full_test`?).

File: cpp.mk (hunk 10 above)
Comment (bossmc) [inline cpp.mk]: `-Wall -Werror`? :grinning:

I think it's about time we switched to `-std=c++11` (seeing as C++14 is already released) it's supported in Trusty's `g++`.

File: cpp.mk (hunk 13 above)
Comment (bossmc) [inline cpp.mk]: I believe that's no longer needed... I'll re-instate it if I'm wrong.

File: cpp.mk (hunk 8 above)
Comment (rkd-msw) [inline cpp.mk]: `make run_test` and `make test`, preserving the existing behaviour?

File: cpp.mk (hunk 11 above)
Comment (bossmc) [inline cpp.mk]: Ah, dark Makefile magic:

```
$ make print-VARIABLE
VARIABLE = <value of VARIABLE>
```

Handy for debugging, I think it's in some (but not all) of our `platform.mk` files.

Comment (rkd-msw) [inline cpp.mk]: Maybe an example as a comment?

File: cpp.mk (hunk 8 above)
Comment (bossmc) [inline cpp.mk]: `run_test` conflicts with the auto-generated `run_<test target>` targets (if the test target is unimaginatively named "test").

Comment (rkd-msw) [inline cpp.mk]: `make test` and `make only_tests`, maybe - I think losing `make test` is a bigger change to adapt to.

Comment (rkd-msw) [inline cpp.mk]: Or, if we'd never run `make run_<test target>` manually, maybe change the auto-generated one to `execute_<test target>`?

Comment (bossmc) [inline cpp.mk]: I currently use `make run_test` quite a bit to not have to wait for valgrind/coverage every time...

Comment (rkd-msw) [inline cpp.mk]: A) That's going to break unless you "happen" to name every test target "test"
B) The situation where `make test` looks plausible but doesn't run coverage or valgrind (which we might not spot), and `make run_test` works as expected (or immediately fails) feels worse than the situation where `make test` works as expected and `make run_test` immediately fails.

(Both of them working as expected is better, though, so I like my `execute_<test target>` suggestion above)

File: cpp.mk (hunk 9 above)
Comment (bossmc) [inline cpp.mk]: Tediously yes, but in the old world, if you ran `make test JUSTTEST="blah"` then you got coverage failures.  Also you're only running the non-valgrind tests twice which doesn't feel too bad to me.

I'll raise an issue to improve this in future.

**Pull request merged**